### PR TITLE
Avoid opening java.io module for tests on JDK17 HZ-687

### PR DESCRIPTION
**Pull request by ldziedziul**
Avoided opening `java.io` module for tests on JDK17

Related to: 
- https://github.com/hazelcast/hazelcast/pull/19896
- https://hazelcast.atlassian.net/browse/HZ-687

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java [hunk 1]
```diff
@@ -1,167 +0,0 @@
...
- * Tests a common compatibility issue: when a (Identified)DataSerializable class first
- * becomes Versioned, it might miss checking input stream for UNKNOWN version (which is
- * the version of an incoming stream from a previous-version member) instead of using
- * in.getVersion.isUnknownOrLessThan(CURRENT).
- */
-@RunWith(PowerMockRunner.class)
-@PowerMockIgnore({"javax.net.ssl.*", "javax.security.*", "javax.management.*"})
-@PrepareForTest(Version.class)
-@Category({QuickTest.class, ParallelJVMTest.class})
-public class VersionedInCurrentVersionTest {
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java:72]: @vbekiaris could you confirm my understanding that this test is no longer needed

**Review by slavagithub — APPROVED** [review]
Looks good to me, thank you!

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java (hunk 1 above)
Comment (ramizdundar) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java:72]: @ldziedziul can you explain why do you think this test is no longer needed?

Comment (ldziedziul) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java:72]: Here's the dump from my conversation with @vbekiaris: 

The reason I'm asking is that in the current form it's impossible to make this test work with JDK17 without opening `java.io` module for tests (which brings some danger of using `java.io` internals in production code).

I've noticed `in.getVersion()` is called only with subsequent `isGreaterOrEqual` and `isGreaterThan` calls

`Version#isUnknownOrLessThan` is called only in
* `RaftService#publishGroupAvailabilityEvents`
* `AbstractDistributedObject#isClusterVersionUnknownOrLessThan`

On the other hand `Version#isLessThan` is used only here:
* `ClusterWideConfigurationService#checkConfigVersion`
* `AbstractDistributedObject#isClusterVersionLessThan`
* `SqlServiceImpl#execute`

I see no places where this test could fail.

From @vbekiaris: 
> The purpose of the test is to avoid allowing someone to write new code using `in.getVersion().isLessThan` in an IDS class’ `readData` that was not `Versioned` in the previous minor version and became `Versioned` in the current codebase.
> can we somehow ensure that and not open `java.io`?

From me:
If we would remove `final` from `Version` then regular (non-powermockito) spy would work

I've also came with another solution: an Archunit rule:

```java
   noClasses().should()
                .callMethod(Version.class, "isLessThan", Version.class)
                .because("newly versioned classes should use 'isUnknownOrLessThan' or similar instead")
                .check(versionedInCurrentVersion);
```

it works but only for direct usage of `in.getVersion().isLessThan` in classes from `versionedInCurrentVersion`. If an input stream is passed to another class then call to `isLessThan` won't be detected. I don't know if it's an issue or not.

To sum it up, if we want to keep this test and avoid `--add-opens` in tests then we have two ways:
- remove `final` from `Version`
- test only direct call using ArchUnit
- something else.... ? e.g. serializing an old class first and then trying to deserialize it with a new class?

What do you think?

Comment (ramizdundar) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java:72]: I believe for the given cases `SerializedObjectsCompatibilityTest` should also fail. So I now believe we can remove this test, if keeping isn't required.

**Review by ramizdundar — APPROVED** [review]
LGTM. I would leave a comment for the package private constructors to show that they are only used for testing.

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java (hunk 1 above)
Comment (vbekiaris) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/VersionedInCurrentVersionTest.java:72]: Good point @ramizdundar , I agree with  you

**Comment by ramizdundar** [comment]
run-lab-run

**Review by sancar — APPROVED** [review]
